src/System.Text.JsonLab/System/Text/Json/JsonWriter.cs

```diff
@@ -3,30 +3,31 @@
 
 using System.Buffers;
 using System.Buffers.Text;
+using System.Buffers.Writer;
 using System.Diagnostics;
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 
 namespace System.Text.JsonLab
 {
-    public struct JsonWriter
+    public ref struct JsonWriter
```

Comment:
Can you help me understand why that would be beneficial?

Also, if we make JsonWriter generic (with the constraint), how can we add support for stream?

Reply:
> Can you help me understand why that would be beneficial?

Two main reasons
1. If you pass `BufferWriter<IBufferWriter<byte>>` a struct formatter; it will box to the `IBufferWriter<byte>` type, allocation and perf issue. Using the generic will avoid the allocation and boxing.
2. The methods called internally to `BufferWriter<T>` (e.g. `GetSpan` and `Advance`) will go via generic interface dispatch in a shared generic; which can't inline and is also the slowest calling convention for the clr. Changing it to the generic means when you pass it a struct type it will use a non-shared generic and direct, inlinable calls.

> Also, if we make JsonWriter generic (with the constraint), how can we add support for stream?

I assume you mean something not currently present?

Make a struct wrapper for Stream and provide a `Create` overload:
```csharp
public struct StreamWriter : IBufferWriter<byte>
{
    private Stream _stream;

    public StreamWriter(Stream stream)
    {
        _stream = stream;
    }

    public void Advance(int count)
    {
        // ...
    }

    public Memory<byte> GetMemory(int sizeHint = 0)
    {
        // ...
    }

    public Span<byte> GetSpan(int sizeHint = 0)
    {
        // ...
    }
}

public static class JsonWriter
{
    public static JsonWriter<TBufferWriter> Create<TBufferWriter>(
        TBufferWriter bufferWriter,
        bool isUtf8,
        bool prettyPrint = false)
    where TBufferWriter : IBufferWriter<byte>
    {
        return new JsonWriter<TBufferWriter>(bufferWriter, isUtf8, prettyPrint);
    }

    // Stream overload
    public static JsonWriter<StreamWriter> Create<TBufferWriter>(
        Stream stream,
        bool isUtf8,
        bool prettyPrint = false)
    where TBufferWriter : IBufferWriter<byte>
    {
        return new JsonWriter<StreamWriter>(new StreamWriter(stream), isUtf8, prettyPrint);
    }
}
```

Reply:
> The methods called internally to BufferWriter<T> (e.g. GetSpan and Advance) will go via generic interface dispatch in a shared generic; which can't inline and is also the slowest calling convention for the clr. Changing it to the generic means when you pass it a struct type it will use a non-shared generic and direct, inlinable calls.

@benaadams is related to CLR behaviour explained here https://blogs.msdn.microsoft.com/joelpob/2004/11/17/clr-generics-and-code-sharing/ ?

Reply:
The `T` parameter is explicitly being specified as an interface: `IBufferWriter<byte>`; so if you create that `BufferWriter<IBufferWriter<byte>>` and give it a struct as the T, then its boxed to the interface and stored in the `BufferWriter<IBufferWriter<byte>>` as the boxed `IBufferWriter<byte>` type

Reply:
> I assume you mean something not currently present?

Yes.

> Make a struct wrapper for Stream and provide a `Create` overload:

That is a great idea! I was trying to see if there was a way to add stream support directly to JsonWriter without exposing additional public types (like the StreamWriter wrapper). However, the boxing and performance benefits you mentioned are definitely compelling.

Reply:
> If you pass `BufferWriter<IBufferWriter<byte>>` a struct formatter; it will box to the `IBufferWriter<byte>` type, allocation and perf issue. Using the generic will avoid the allocation and boxing.

I wasn't seeing any allocations because the formatter used within the benchmarks is a class (ArrayFormatter). Good point.

Reply:
Ah, missed that. Can always measure the effects of the call costs by creating a second struct that just wraps and using that in a test instead of `ArrayFormatter` e.g.
```csharp
public struct ArrayFormatterWrapper : ITextBufferWriter, IDisposable
{
    private ArrayFormatter _arrayFormatter;

    public ArrayFormatter(int capacity, SymbolTable symbolTable, ArrayPool<byte> pool = null)
    {
        _arrayFormatter = new ArrayFormatterInternal(capacity, symbolTable, pool);
    }

    public int CommitedByteCount => _arrayFormatter.CommitedByteCount;

    public void Clear()
    {
        _arrayFormatter.Clear();
    }

    public ArraySegment<byte> Free => _arrayFormatter.Free;
    public ArraySegment<byte> Formatted => _arrayFormatter.Formatted;

    public SymbolTable SymbolTable => _arrayFormatter.SymbolTable;

    public Memory<byte> GetMemory(int minimumLength = 0)
    {
        _arrayFormatter.GetMemory(minimumLength);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public Span<byte> GetSpan(int minimumLength = 0)
    {
        _arrayFormatter.GetSpan(minimumLength);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void Advance(int bytes)
    {
        _arrayFormatter.Advance(bytes);
    }
    public int MaxBufferSize { get; } = _arrayFormatter.MaxBufferSize;

    public void Dispose()
    {
        _arrayFormatter.Dispose();
    }
}
```